Update control, description, and gazebo packages for rviz/moveit control

acmiyaguchi opened the pull request:
This configures the `finalarm_description_position_control` urdf to be controllable via the `finalarm_gazebo_trajectory_control` launch file. I've tested that this works correctly on Ubuntu 20.04/ROS Noetic/Gazebo 11.

```bash
roslaunch finalarm_gazebo finalarm_world_t.launch
roslaunch finalarm_moveit_config move_group.launc
roslaunch finalarm_moveit_config moveit_rviz.launch
```

This aligns with my expectations of running Handy in the lab. The gazebo configuration wasn't publishing to the correct topics, and was using extra namespaces. I took [notes of topics on Mary](https://github.gatech.edu/VisMan/Mary_ROS/blob/master/notes.md) when I was in the lab, so we expect topics in the following shape:

```
/mary_joint_trajectory_action_controller/command
/mary_joint_trajectory_action_controller/follow_joint_trajectory/cancel
/mary_joint_trajectory_action_controller/follow_joint_trajectory/feedback
/mary_joint_trajectory_action_controller/follow_joint_trajectory/goal
/mary_joint_trajectory_action_controller/follow_joint_trajectory/result
/mary_joint_trajectory_action_controller/follow_joint_trajectory/status
/mary_joint_trajectory_action_controller/state
/mary_position_controller_1/command
/mary_position_controller_1/state
/mary_position_controller_2/command
/mary_position_controller_2/state
/mary_position_controller_3/command
/mary_position_controller_3/state
/mary_position_controller_4/command
/mary_position_controller_4/state
/mary_position_controller_5/command
/mary_position_controller_5/state
/mary_position_controller_6/command
/mary_position_controller_6/state
```

Indeed, we get the following topics:

```bash
/finalarm_joint_trajectory_action_controller/command
/finalarm_joint_trajectory_action_controller/follow_joint_trajectory/cancel
/finalarm_joint_trajectory_action_controller/follow_joint_trajectory/feedback
/finalarm_joint_trajectory_action_controller/follow_joint_trajectory/goal
/finalarm_joint_trajectory_action_controller/follow_joint_trajectory/result
/finalarm_joint_trajectory_action_controller/follow_joint_trajectory/status
/finalarm_joint_trajectory_action_controller/state
/finalarm_position_controller_8/command
/finalarm_position_controller_9/command
```

There are a couple of things to note:

- The position controller launch file seems to work, but we can't have both position controllers and a trajectory controller at the same time.
- I have not updated the torque controller file with any of the changes here. I don't understand why we have this configuration, and why the PID parameters differ.
- The urdf for the base handy is broken, probably because it isn't setup to run in gazebo. I did a diff of the position controller and the torque controller, and they seem to be the same aside from a few parameters. I did not update the torque controller with the new position controller definitions.
- I did not touch the simplified files. I don't understand what these are for.

@ruinianxu Can you advise on what changes I should make here to make the code consistent? Should I apply my changes to all of the files and test that they all work as expected? Is there code that is no longer used that we can completely get rid of here?

acmiyaguchi commented on ros/finalarm_control/launch/finalarm_gazebo_position_torque_control.launch:
```diff
@@ -18,7 +18,7 @@
   <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
     respawn="false" output="screen">
-    <remap from="/joint_states" to="/finalarm/joint_states" />
+    <remap from="/joint_states" to="/finalarm_manager/joint_states" />
```
Oops, this needs to get fixed in the same way that the position launch file was fixed; this was me figuring out how namespaces worked.

acmiyaguchi commented on ros/finalarm_control/config/finalarm_gazebo_trajectory_control.yaml:
```diff
@@ -1,28 +1,27 @@
-finalarm:
```
We remove this namespace so we publish to the main topic, like the physical robot does. Namespacing the topics does make sense though generally.

ruinianxu commented:
@acmiyaguchi 
I forgot why I implemented a torque controller but you can get rid of all torque-related code since they were never really used. I will take a look at the PR today and make it merge.

ruinianxu commented:
@acmiyaguchi 
I did a quick review of code and have some my thoughts. Correct me if I am wrong. I think I understand why you said you got some issues with MoveIt!. The basic reason is that Handy MoveIt! package was configured based on the old ROS (i.e., Kinetic). After we upgrade to Noetic, we will need to re-generate Handy MoveIt package. I think we don't need these changes and the cleanest way is to confiugre a noetic-version finalarm_moveit_config and maintain a new branch named noetic-devel. 

For MoveIt! with gazebo, I mean something like https://www.youtube.com/watch?v=Q5ppJhRfGKo. So far, we will have to send commands to individual motors to form a trajectory. The ideal way is to plan a trajectory and send it to trajectory controller.

acmiyaguchi commented:
> The basic reason is that Handy MoveIt! package was configured based on the old ROS (i.e., Kinetic).

Actually, this is incorrect. The current moveit package works fine both with Gazebo and the physical robot. The reason why there are issues is because the simulated position/trajectory controllers are publishing to the incorrect topics, and that there were missing dependencies. Note that most of my changes were in `finalarm_control` and `finalarm_description`.

I will get rid of the position-torque related code. I may also make the `finalarm_description_position_control.urdf` the only model that we keep in the repository, because the other two are copies in some form (without the gazebo components).

Can we get rid of the `finalarm_simplified` packages? This looks like pure copy-and-paste overhead in terms of development. Is there any case we'd want to run the robot without the end effector?

ruinianxu commented:
> > The basic reason is that Handy MoveIt! package was configured based on the old ROS (i.e., Kinetic).
> 
> Actually, this is incorrect. The current moveit package works fine both with Gazebo and the physical robot. The reason why there are issues is because the simulated position/trajectory controllers are publishing to the incorrect topics, and that there were missing dependencies. Note that most of my changes were in `finalarm_control` and `finalarm_description`.
> 
> I will get rid of the position-torque related code. I may also make the `finalarm_description_position_control.urdf` the only model that we keep in the repository, because the other two are copies in some form (without the gazebo components).
> 
> Can we get rid of the `finalarm_simplified` packages? This looks like pure copy-and-paste overhead in terms of development. Is there any case we'd want to run the robot without the end effector?

I see.  finalarm_simplified was the package created by the request of Dr. Vela. He wanted a version with simple geometry. The only difference is in the mesh file. For simplified arm, the mesh for each link is just cube.

ruinianxu commented:
> > The basic reason is that Handy MoveIt! package was configured based on the old ROS (i.e., Kinetic).
> 
> Actually, this is incorrect. The current moveit package works fine both with Gazebo and the physical robot. The reason why there are issues is because the simulated position/trajectory controllers are publishing to the incorrect topics, and that there were missing dependencies. Note that most of my changes were in `finalarm_control` and `finalarm_description`.
> 
> I will get rid of the position-torque related code. I may also make the `finalarm_description_position_control.urdf` the only model that we keep in the repository, because the other two are copies in some form (without the gazebo components).
> 
> Can we get rid of the `finalarm_simplified` packages? This looks like pure copy-and-paste overhead in terms of development. Is there any case we'd want to run the robot without the end effector?

For `finalarm_description_position_control.urdf`, you can separate lines 630-798 out into a xacro file. Then we can set an input argument to decide whether to load xacro file when we upload urdf file into parameter server. This is the way how kinova does.

acmiyaguchi commented:
> I see. finalarm_simplified was the package created by the request of Dr. Vela. He wanted a version with simple geometry. The only difference is in the mesh file. For simplified arm, the mesh for each link is just cube.

Fair enough, I'll make sure to clean up these files, too. It looks like the simplified package uses xacro and includes a kinect in the world; it should be straightforward to modify and test this alongside other changes.

acmiyaguchi commented on ros/finalarm_description/CMakeLists.txt:
```diff
@@ -2,12 +2,16 @@ cmake_minimum_required(VERSION 2.8.3)
 
+# build urdf files into the share directory for backwards compatibility
+file(GLOB xacro_files urdf/*.urdf.xacro)
+xacro_add_files(${xacro_files} INSTALL DESTINATION urdf)
```
We can build xacro files so we don't have to run xacro at runtime. This helps maintain compatibility with existing moveit config.

acmiyaguchi commented on ros/finalarm_description/launch/gazebo.launch:
```diff
@@ -1,20 +1,31 @@
+  />
+  <param
+    name="robot_description"
+    command="$(find xacro)/xacro $(arg model) use_gazebo:=true"
```
I've defaulted to using xacro where-ever possible. There's some backwards compatibility baked in, but if it's broken on other code, they should instead pin the repository to an older version or upgrade.

acmiyaguchi commented on ros/finalarm_simplified_control/CMakeLists.txt:
```diff
@@ -9,11 +9,3 @@ find_package(roslaunch)
 
 install(DIRECTORY launch DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
 install(DIRECTORY config DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
-install(DIRECTORY script DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
```
I removed these scripts from the finalarm_simplified_control and delegate them to the finalarm_control. I dont think we should duplicate the scripts across the modules.

acmiyaguchi commented on ros/finalarm_simplified_control/launch/controller_manager.launch:
```diff
@@ -1,14 +1,3 @@
-                update_rate: 100
-    </rosparam>
-  </node>
+  <include file="$(find finalarm_control)/launch/controller_manager.launch" />
```
Did some refactoring here to reference the finalarm_control files.

acmiyaguchi commented on ros/finalarm_simplified_control/launch/finalarm_gazebo_control.launch:
```diff
@@ -1,22 +1,20 @@
+    <arg
+      name="spawner_args"
+      value="
+        finalarm_position_controller_1
```
The only thing about these is that they don't follow the convention of using `finalarm_simplified` as the model name. We don't use namespacing of the topics anywhere else, so this would not work in gazebo and on the real robot the same way. Arguable how this should be implemented.

acmiyaguchi commented on ros/finalarm_description/urdf/finalarm_description.urdf.xacro:
```diff
@@ -0,0 +1,504 @@
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="finalarm_description">
+  <xacro:arg name="use_gazebo" default="false" />
+
+  <xacro:if value="$(arg use_gazebo)">
```
Set this up so this file contains the source of truth for the robot. Added a few nice macros to the end here. Not sure if it's as easy as swapping out the STL files for the actual description, but it's worth considering parameterizing this further so we can reuse it for the simplified robot.

acmiyaguchi commented on ros/finalarm_description/urdf/finalarm_description.urdf.xacro:
```diff
@@ -0,0 +1,504 @@
+        <turnGravityOff>False</turnGravityOff>
+        <selfCollide>False</selfCollide>
+      </gazebo>
+    </xacro:macro>
```
The macros ended up being pretty nice here, though, though I wanted to use loops of some kind.

acmiyaguchi reviewed (commented):
@ruinianxu I went through and did a lot more refactoring. That addresses some of the comments from before. I haven't tested gazebo-rviz functionality, but I plan on adding a new launchfile in both gazebo packages to make it easy to do.

The simplified handy differs a bit in terms of the structure. Also, it might be worth considering generating the urdf at build time from the xacro urdf, and maybe sharing a number of components from the base handy. No idea if this is worth the time to do, since it's currently functional.

I'd like wind down on these changes though, so let me know if there's anything that seems completely out of place.

acmiyaguchi commented:
Alright, I added in a bunch of manual testing and it looks like all these commands should run okay:

```bash
# check for urdf loading
roslaunch finalarm_description display.launch
# check robot position topics
roslaunch finalarm_description gazebo.launch
roslaunch finalarm_gazebo robot_world.launch
roslaunch finalarm_gazebo robot_world.launch control_mode:=position
# check operation of robot
roslaunch finalarm_moveit_config demo.launch
# check gazebo and moveit operate together
roslaunch finalarm_gazebo robot_world_moveit.launch

roslaunch finalarm_simplified_description display.launch
roslaunch finalarm_simplified_description gazebo.launch
roslaunch finalarm_simplified_gazebo robot_world.launch
roslaunch finalarm_simplified_gazebo robot_world_urdf.launch
roslaunch finalarm_simplified_gazebo robot_world_xacro.launch
roslaunch finalarm_simplified_moveit_config demo.launch
roslaunch finalarm_simplified_gazebo robot_world_moveit.launch
```

This needs to be tested against the actual robot. I made a few more decisions here:

- The simplified robot now depends on the control module of the full robot. It publishes to the same topics (because I couldn't figure out a good way to remap the trajectory action controller via the robot state publisher).
- I removed finalarm out of the launchfile names -- the namespace of the package should take care of most of this.
- I've fixed the python scripts so they can run against the simulated controllers; this means that we simply spawn the robot, run the state publisher, and we should be in the same state as the physical robot.

ruinianxu commented:
@acmiyaguchi Thank you so much for all your work. I think this PR is ready to go except one old issue. It pops out `ImportError: No module named yaml` when running `disable_torque_handy.py`. Do you have a solution for it?

acmiyaguchi commented:
Looks like we need to add https://index.ros.org/d/python-yaml/ to the list of dependencies (and make sure to run the rosdep install to pick up the new system packages). I also found a small bug while testing this code with one of the other students in the autobots group. I'll make these changes later tonight.

ruinianxu commented:
> Looks like we need to add https://index.ros.org/d/python-yaml/ to the list of dependencies (and make sure to run the rosdep install to pick up the new system packages). I also found a small bug while testing this code with one of the other students in the autobots group. I'll make these changes later tonight.

I was still getting the same error after the commit. Then I followed this [link](https://blog.csdn.net/mail_/article/details/126933646?spm=1001.2101.3001.6650.6&utm_medium=distribute.pc_relevant.none-task-blog-2%7Edefault%7EBlogCommendFromBaidu%7ERate-6-126933646-blog-121678556.235%5Ev27%5Epc_relevant_default_base1&depth_1-utm_source=distribute.pc_relevant.none-task-blog-2%7Edefault%7EBlogCommendFromBaidu%7ERate-6-126933646-blog-121678556.235%5Ev27%5Epc_relevant_default_base1&utm_relevant_index=13) to addres the isse of missing yaml. However I got the error of no rospkg after fixing yaml. It seemed it is a common issue for missing these two packages. I eventually fixed the error of no rospkg by this [link](https://answers.ros.org/question/245967/importerror-no-module-named-rospkg-python3-solved/) .

I think this PR is ready to go. The above two errors have no relation to the PR.
